File: yangwanhong-exam/src/test/java/com/example/demo/ControllerTest.java
package com.example.demo;

import com.example.demo.controller.UrlConvertController;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultHandlers;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

@SpringBootTest
@RunWith(SpringRunner.class)
public class ControllerTest {

	@Autowired
	private UrlConvertController urlConvertController;

	private MockMvc mockMvc;

	@Before
	public void setup() {
		mockMvc = MockMvcBuilders.standaloneSetup(urlConvertController).build();
	}

	@Test
	public void getShortUrlTest() throws Exception {

		MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get("/urlConvert/getShortUrl?url=https://www.baidu.com/fasf"))
				.andExpect(MockMvcResultMatchers.status().isOk())
				.andDo(MockMvcResultHandlers.print())
				.andReturn();

		Assert.assertTrue("短域名超出长度限制",mvcResult.getResponse().getContentAsString().length()<=8);

	}

	@Test
	public void convertToLongUrl() throws Exception {
		MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get("/urlConvert/convertToLongUrl?url=https://www.baidu.com/fasf"))
				.andExpect(MockMvcResultMatchers.status().isOk())
				.andDo(MockMvcResultHandlers.print())
				.andReturn();
	}

	@Test
	public void convertShortAndLong() throws Exception {
		String longUrl = "ggafasv123f4Gfasf";
		MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get("/urlConvert/getShortUrl?url="+longUrl))
				.andExpect(MockMvcResultMatchers.status().isOk())
				.andDo(MockMvcResultHandlers.print())
				.andReturn();
		String shortUrl = mvcResult.getResponse().getContentAsString();

		mvcResult = mockMvc.perform(MockMvcRequestBuilders.get("/urlConvert/convertToLongUrl?url="+shortUrl))
				.andExpect(MockMvcResultMatchers.status().isOk())
				.andDo(MockMvcResultHandlers.print())
				.andReturn();

		System.out.println(mvcResult.getResponse().getContentAsString());
		Assert.assertTrue("转换错误",mvcResult.getResponse().getContentAsString().equals(longUrl));

	}
}

File: java/src/test/java/com/example/urlshortener/ControllerTest.kt
package com.example.urlshortener

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.boot.web.server.LocalServerPort
import org.springframework.http.MediaType
import org.springframework.test.annotation.DirtiesContext
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext
import  org.hamcrest.Matchers.containsString;
import  org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import  org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import  org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import  org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import  org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc
class ShortUrlControllerTest {

    companion object {
        private const val BASE_PATH = "/api/v1"
    }

    @Autowired
    private val mockMvc: MockMvc? = null

    @Test
    @Throws(Exception::class)
    fun testShortenUrl() {
        val input = "http://fakercorp.com/fake-long-path-with-many/"
        mockMvc!!.perform(
            post("$BASE_PATH/shorten")
                .contentType(MediaType.APPLICATION_JSON).content(input)
        ).andDo(print())
            .andExpect(status().isOk)
            .andExpect(content().string(containsString("http://localhost$BASE_PATH/shorten/C6IHIY")))
            .andReturn()
    }

//    @Test
//    @Throws(java.lang.Exception::class)
//    fun testFindUrl() {
//        val id = "C6IHIY"
//        val longUrl = "http://fakercorp.com/fake-long-path-with-many/"
//        mockMvc!!
//            .perform(get("$BASE_PATH/query/$id"))
//            .andDo(print())
//            .andExpect(status().isOk)
//            .andExpect(content().string(containsString("C6IHIY")))
//    }
}